tiledb/type/range/range.h

```diff
@@ -451,42 +452,52 @@ Status check_range_is_subset(const Range& superset, const Range& range) {
  *
  * @param range The range to check.
  */
-template <
-    typename T,
-    typename std::enable_if_t<
-        std::is_arithmetic_v<T> || std::is_same_v<T, std::string_view>>* =
-        nullptr>
-void check_range_is_valid(const Range& range) {
+template <class T>
+void check_range_is_valid(const Range& range)
+  requires(std::is_arithmetic_v<T>)
+{
   // Check has data.
   if (range.empty())
     throw std::invalid_argument("Range is empty");
 
-  // Compare string views
-  if constexpr (std::is_same_v<T, std::string_view>) {
-    auto start = range.start_str();
-    auto end = range.end_str();
-    // Check range bounds
-    if (start > end)
-      throw std::invalid_argument(
-          "Lower range bound " + std::string(start) +
-          " cannot be larger than the higher bound " + std::string(end));
-  } else {
-    if (range.size() != 2 * sizeof(T))
-      throw std::invalid_argument(
-          "Range size " + std::to_string(range.size()) +
-          " does not match the expected size " + std::to_string(2 * sizeof(T)));
-    auto r = (const T*)range.data();
-    // Check for NaN
-    if constexpr (std::is_floating_point_v<T>) {
-      if (std::isnan(r[0]) || std::isnan(r[1]))
-        throw std::invalid_argument("Range contains NaN");
-    }
-    // Check range bounds
-    if (r[0] > r[1])
-      throw std::invalid_argument(
-          "Lower range bound " + std::to_string(r[0]) +
-          " cannot be larger than the higher bound " + std::to_string(r[1]));
-  }
+  if (range.size() != 2 * sizeof(T))
+    throw std::invalid_argument(
+        "Range size " + std::to_string(range.size()) +
+        " does not match the expected size " + std::to_string(2 * sizeof(T)));
+  auto r = (const T*)range.data();
+  // Check for NaN
+  if constexpr (std::is_floating_point_v<T>) {
+    if (std::isnan(r[0]) || std::isnan(r[1]))
+      throw std::invalid_argument("Range contains NaN");
+  }
+  // Check range bounds
+  if (r[0] > r[1])
+    throw std::invalid_argument(
+        "Lower range bound " + std::to_string(r[0]) +
+        " cannot be larger than the higher bound " + std::to_string(r[1]));
+};
+
+/**
+ * Performs correctness checks for a valid range. If any validity checks fail
+ * an exception is thrown.
+ *
+ * @param range The range to check.
+ */
+template <class T>
+void check_range_is_valid(const Range& range)
+  requires(std::same_as<T, std::string_view>)
```

Comment:
Specialization syntax would be clearer here. There's only one possible class to bind to.

Reply:
We cannot add a specialization on the previous `check_range_is_valid` because its `requires` clause does not allow for `string_view` (I tried it and it failed). I can't think of another way because `T` does not appear anywhere in the signature.

Reply:
It's better for compile-time error handling to declare a (non-specialized) template without any requirement and then add narrower specializations on top of it. C++20 does not require that non-specialized template when `requires` is present, but it's often a good idea to add it in anyway. That way if the template is invoked without a valid type argument at least the programmer knows that the template is in scope. In addition, the non-specialized template can be defined with `static_assert(false)` in its body to provide a very explicit error message.

In order to add an ordinary specialization, you need at least a non-specialized template declaration.